packages/x-data-grid/src/DataGrid/useDataGridProps.ts

```diff
@@ -108,6 +108,7 @@ export const useDataGridProps = <R extends GridValidRowModel>(inProps: DataGridP
     () => ({
       ...DATA_GRID_PROPS_DEFAULT_VALUES,
       ...themedProps,
+      paginationMode: themedProps.paginationMode ?? DATA_GRID_PROPS_DEFAULT_VALUES.paginationMode,
```

Comment:
when `pagination={undefined}` is set, `pagintionMode` key will be present in `themedProps` and overrides `paginationMode` from `DATA_GRID_PROPS_DEFAULT_VALUES` but when `pagination` prop is not explicitly passed, then value of `paginationMode` from `DATA_GRID_PROPS_DEFAULT_VALUES` will be considered since `themedProps` won't have `paginationMode` key.

With below change, both not passing `pagintionMode` and `pagintionMode={undefined}` will behave same

paginationMode: `themedProps.paginationMode ?? DATA_GRID_PROPS_DEFAULT_VALUES.paginationMode`

Reply:
This won't solve the problem for other props so I'm not sure it's the right way of doing it :grimacing:

Reply:
I see, what do you think of below approach

```
  const injectDefaultProps = React.useMemo(() => {
    return Object.keys(DATA_GRID_PROPS_DEFAULT_VALUES).reduce((acc, key) => {
      acc[key] =
        themedProps[key as keyof DataGridProps<R>] ??
        DATA_GRID_PROPS_DEFAULT_VALUES[key as keyof DataGridPropsWithDefaultValues<any>];
      return acc;
    }, {} as DataGridPropsWithDefaultValues<any>);
  }, [themedProps]);

  return React.useMemo<DataGridProcessedProps<R>>(
    () => ({
      ...themedProps,
      ...injectDefaultProps,
      localeText,
      slots,
      ...DATA_GRID_FORCED_PROPS,
    }),
    [themedProps, localeText, slots, injectDefaultProps],
  );
```

Reply:
Yeah we probably need something like that
I'll let the team working on the grid review the exact implementation.

By the way I wouldn't be surprised if we had similar issues on other packages such as `@mui/x-date-pickers`

Reply:
Reminds me a bit of lodash [`defaults`](https://lodash.com/docs/4.17.15#defaults) as well.

Reply:
refactored logic from just fixing `paginationMode={undefined}` to fixing all defaultProps. commit: https://github.com/mui/mui-x/pull/13349/commits/edb438f0fee45b81bf113267393bba8a93c60c08